### DISCUSSION TITLE
Why is this urllib3 annotation not accepted by `--verifytypes`?

### DISCUSSION THREAD

**Discussion by Pyprohly** [microsoft/pyright#2484] [Q&A] [answered]
Consider the following module in a library:

```python
import urllib3
a: urllib3.poolmanager.PoolManager

import requests
b: requests.Session
```

Pyright complains: “error: Type unknown for variable "pkg.mod1.a"”.

I can’t figure out what is different about the urllib3 annotation that makes the type completeness checker not like it?

**Comment by erictraut** [comment, marked as answer]
The `urllib3` library is not a "py.typed" library, and it doesn't ship with type stubs. If you are importing `PoolManager` from `urllib3` and exposing the type as part of your "py.typed" library's interface, the type of `PoolManager` will be Unknown. To fix this, you would need to work with the maintainers of `urllib3` to make their library "py.typed" as well.

If you want to focus on verifying the type completeness of the the symbols that are directly within your control (as opposed to those that your library imports from other libraries), you can use the `--ignoreexternal` option along with `--verifytypes`.

**Reply by Pyprohly** [reply]
Sorry I should have specified, I am using the `--ignoreexternal` option. There’d be too many errors I couldn’t fix if I didn’t use the `--ignoreexternal` option because I’m trying to wrap multiple HTTP libraries, including [requests], [urllib3], [httpx], and [aiohttp]. And only urllib3’s ‘session’ class is producing an error and I don’t understand why.

[requests]: https://github.com/psf/requests
[urllib3]: https://github.com/urllib3/urllib3
[httpx]: https://github.com/encode/httpx
[aiohttp]: https://github.com/aio-libs/aiohttp

How does one tell if a package is py.typed? As far as I can see there is no py.typed file in the requests package either but the `requests.Session` annotation doesn’t produce an error like urllib3’s `PoolManager` does?

**Reply by erictraut** [reply]
The `requests` library is [included in typeshed](https://github.com/python/typeshed/tree/master/stubs/requests), so it has complete type information provided in the form of type stubs. Pyright ships with all typeshed stubs, so the type information for `requests` is effectively "built in" to pyright.

In contrast, the `urllib3` library is not included in typeshed, and it doesn't ship with type stubs, nor does it have inlined type information.

You can tell whether a library is "py.typed" by looking for an empty text file in the root directory of the package called "py.typed". For details about this file, refer to [PEP 561](https://www.python.org/dev/peps/pep-0561/).

Is your library available in a public github repo? I'd like to understand how you're using `PoolManager` within your library's public API. If you are simply importing it and using it within a type annotation for an input parameter, function return type, or instance variable, it shouldn't be flagged as an unknown type if you use `--ignoreexternal`. Perhaps you're using `PoolManager` as the base class for one of your classes?

**Reply by Pyprohly** [reply]
I did not consider that Pyright shipped with a copy of typeshed, otherwise I would have used a different HTTP library that isn’t in typeshed as an example.

The issue is easily reproducible by creating a dummy library and simply using the `PoolManager` annotation on some public symbol.

My repo isn’t public but I can make it public now. https://github.com/Pyprohly/redditwarp (`pip install redditwarp`)

I’m using `PoolManager` only as a plain annotation and not in any other special way. The line that is flagged is [here]. The other HTTP transport adaptor modules in my library follow a similar structure and they have no problems.

[here]: https://github.com/Pyprohly/redditwarp/blob/7bb9d59e80814edc44223d3274bf87e06b41936e/redditwarp/http/transport/urllib3.py#L21

The same issue occurs with the websocket-client lib, as you’ll see from the Pyright output.

**Reply by erictraut** [reply]
Thanks, that was very helpful. I was able to identify and fix the problem. It will be addressed in the next release of pyright.
